core/modules/block/block.module

```diff
@@ -279,6 +279,19 @@ function block_custom_block_save(array $edit, $delta = NULL) {
   $config->setData($block);
   $config->save();
 
+  // Save each file as permanent, preventing it from being deleted.
+  $fids = filter_parse_file_fids($block['body']['value']);
+  $files = file_load_multiple($fids);
+  foreach ($files as $fid => $file) {
+    if ($file && $file->status !== FILE_STATUS_PERMANENT) {
+      $file->status = FILE_STATUS_PERMANENT;
+      $file->save();
+    }
+    // Essentially this claims that the file is "self-referencing", which means
+    // it will never be deleted until a file management tool is available.
+    file_usage_add($file, 'file', 'file', $file->fid);
+  }
+
```

Comment:
So this is because you can create blocks under Admin -> Structure -> Custom blocks. You could create a block but not place it within a layout. There is duplication and this code will fire twice when creating a block in the Layout UI, but as the files will only be marked permanent once, it won't have an impact on the file usage entries.

However, there's definitely got to be a smarter way around this. I'll make a centralized function for this task.